### PR TITLE
Material Date Picker redesign

### LINKED ISSUE #50133: Material design component: mobile date input picker
## Use case

When a user needs to input a date in the distance past or future (such as a birth date), the material design guidelines recommend using a mobile date input picker. The `showDatePicker()` method in flutter does not support this type of picker.

image

I cannot find a flutter library implementing this.

## Proposal

Add a material widget implementing the mobile date input picker show in the material design guidelines: https://material.io/components/pickers/#mobile-pickers.

### PR DESCRIPTION
**Pull request by darrenaustin**
## Description

The design for the Material Date Picker has been updated:

https://material.io/components/pickers/#mobile-pickers

This PR updates Flutter's implementation to match the new design.

DatePickerM2

In addition to just updating the visuals and controls, a new text input mode was added to the dialog as per the design.

The API changes are documented at: https://flutter.dev/go/material-date-picker-redesign

Note that this does not include support for the Date Range picker as that will be coming in a future PR.

Also, credit to @rami-a who implemented a similar internal design that I based this on. However, any problems were introduced by me, so don't blame him :).

## Related Issues

* Fixes #50133

## Tests

The tests in `date_picker_test.dart` have been updated to test the new UI implementation. Tests for the new text input mode were also added to cover that.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/material/date_picker_test.dart [hunk 1]
```diff
@@ -150,672 +70,850 @@ void _tests() {
...
-    firstDate = DateTime(2017, DateTime.january, 10);
-    lastDate = DateTime(2017, DateTime.january, 20);
-    selectableDayPredicate = (DateTime day) => day.day.isEven;
-    await preparePicker(tester, (Future<DateTime> date) async {
-      await tester.tap(find.text('10')); // Even, works.
-      await tester.tap(find.text('13')); // Odd, doesn't work.
-      await tester.tap(find.text('17')); // Odd, doesn't work.
-      await tester.tap(find.text('OK'));
-      expect(await date, equals(DateTime(2017, DateTime.january, 10)));
+    testWidgets('Can select initial date picker mode', (WidgetTester tester) async {
```
Comment (rami-a) [inline packages/flutter/test/material/date_picker_test.dart]: This test appears to be a duplicate of an earlier one in here

**Review by rami-a — COMMENTED** [review]
Looks mostly good, comments are pretty minor but noticed a couple of small bugs when testing it out.

- Header overflows in landscape mode with text input and keyboard up
- Landscape mode should probably cap the width (especially in text input mode), otherwise it looks very wide.
- Rotating the device does not maintain the current month the user is viewing, it goes back to the initial month.

File: packages/flutter/lib/src/material/pickers/date_utils.dart [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+
+// Common helper functions for all pickers.
+
+import '../material_localizations.dart';
+
+/// Returns a [DateTime] with just the date of the original, but no time set.
+DateTime dateOnly(DateTime date) {
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_utils.dart]: I thought about that, but I am bit ambivalent about extension methods. They do make things easier to read, but it is not obvious that they aren't part of the main object's API, which can lead to confusion. Given that these are all internal anyway, I just left them as util functions.

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+      return Container(
+        padding: const EdgeInsets.symmetric(horizontal: 24),
+        height: orientation == Orientation.portrait ? _inputPortraitHeight  : _inputLandscapeHeight,
+        child: Column(
+          children: <Widget>[
+            SizedBox(height: orientation == Orientation.portrait ? 16 : 26),
+            TextFormField(
+              decoration: InputDecoration(
+                border: const UnderlineInputBorder(),
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/input_date_picker.dart]: My hope is that we can expose these kinds of things in a DatePickerTheme pass.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+                    ],
+                  ),
+                ),
+              ],
+            );
+        }
+        return null;
+      }),
+      insetPadding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 24.0),
+      shape: const RoundedRectangleBorder(
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: Yup. My thought was to update the DialogTheme to match the rest of the current theme patterns, in which case we could update the defaults and then we wouldn't need to hard code those here (or just have these be fall backs when the DialogTheme didn't override them).

**Comment by darrenaustin** [comment]
@rami-a, thanks for the review. I think I have either addressed all your feedback. There are some issues, like customizability which we will deal with in a DatePickerTheme pass in another PR.

> * Header overflows in landscape mode with text input and keyboard up
> * Landscape mode should probably cap the width (especially in text input mode), otherwise it looks very wide.
> * Rotating the device does not maintain the current month the user is viewing, it goes back to the initial month.

These should all be fixed with my latest commits. Please let me know if you see anything else.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 5]
```diff
@@ -434,6 +475,23 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
     'December',
   ];
 
+  /// Returns the number of days in a month, according to the proleptic
+  /// Gregorian calendar.
+  ///
+  /// This applies the leap year logic introduced by the Gregorian reforms of
+  /// 1582. It will not give valid results for dates prior to that time.
+  int _getDaysInMonth(int year, int month) {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material_localizations.dart]: It's too bad this isn't already part of DateTime.  Maybe propose a PR for the Dart team?

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 6]
```diff
@@ -467,6 +525,21 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
   @override
   String formatYear(DateTime date) => date.year.toString();
 
+  @override
+  String formatCompactDate(DateTime date) {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material_localizations.dart]: So what happens in other locales?  Is this just a fallback, or are the other locales just not implemented yet?

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 7]
```diff
@@ -0,0 +1,1050 @@
...
+    final Widget picker = Flexible(
+      child: SingleChildScrollView(
+        child: SizedBox(
+          height: _maxDayPickerHeight,
+          child: _buildPicker(),
+        ),
+      ),
+    );
+
+    final Widget dialog = OrientationBuilder(
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Why not just return the `OrientationBuilder` here and not bother creating `dialog` at all?

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 8]
```diff
@@ -0,0 +1,1050 @@
...
+  final ValueChanged<DateTime> onChanged;
+
+  @override
+  _YearPickerState createState() => _YearPickerState();
+}
+
+class _YearPickerState extends State<_YearPicker> {
+  ScrollController scrollController;
+
+  // The approximate number of years necessary to fill the available space.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Can you explain this a little more?  Why is it approximate, not exact, and what available space is it filling?

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 9]
```diff
@@ -0,0 +1,1050 @@
...
+  static const int minYears = 18;
+
+  @override
+  void initState() {
+    super.initState();
+
+    // Set the scroll position to approximately center the initial year.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Again, why approximate, and when does the approximation get resolved? Is there some kind of step that narrows down the value to the exact value?

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+/// and [useRootNavigator] must be non-null.
+///
+/// The [builder] parameter can be used to wrap the dialog widget
+/// to add inherited widgets like [Theme].
+///
+/// An optional [initialDatePickerMode] argument can be used to have the
+/// calendar date picker initially appear in the [DatePickerMode.year] or
+/// [DatePickerMode.day] mode. It defaults to [DatePickerMode.day], and
+/// must be non-null.
+///
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: Extra line can be removed.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+import 'date_utils.dart' as utils;
+import 'input_date_picker.dart';
+
+const Size _calendarPortraitDialogSize = Size(330.0, 518.0);
+const Size _calendarLandscapeDialogSize = Size(496.0, 346.0);
+const Size _inputPortraitDialogSize = Size(330.0, 270.0);
+const Size _inputLandscapeDialogSize = Size(496, 160.0);
+const Duration _dialogSizeAnimationDuration = Duration(milliseconds: 200);
+
+/// Shows a dialog containing a Material Design date picker.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: It would be nice to add an interactive sample here sometime (not in this PR).

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+    Navigator.pop(context);
+  }
+
+  void _handelEntryModeToggle() {
+    setState(() {
+      switch (_entryMode) {
+        case DatePickerEntryMode.calendar:
+          _entryMode = DatePickerEntryMode.input;
+          break;
+
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: nit: don't need the extra line here, IMO.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 13]
```diff
@@ -0,0 +1,1050 @@
...
+const double _yearPickerRowHeight = 52.0;
+const double _yearPickerRowSpacing = 8.0;
+
+/// Displays a grid of days for a given month and allows the user to select a date.
+///
+/// Days are arranged in a rectangular grid with one column for each day of the
+/// week. It provides controls to change the year and month that the grid is showing.
+///
+/// The calendar picker widget is rarely used directly. Instead, consider using
+/// [showDatePicker], which will create a dialog that uses this as well as provides
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: which creates a Dialog that contains a [CalendarDatePicker] and provides an optional compact view where the user can enter a date as a line of text.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 14]
```diff
@@ -0,0 +1,1050 @@
...
+    }
+  }
+
+  void _setCurrentMonth(DateTime date) {
+    setState(() {
+      if (_currentDisplayedMonthDate.year != date.year || _currentDisplayedMonthDate.month != date.month) {
+        _currentDisplayedMonthDate = DateTime(date.year, date.month);
+        _previousMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, -1);
+        _nextMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, 1);
+        _pageController?.dispose();
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: This seems a bit risky. Presumably there's no way for the now-disposed page controller to be accessed, even though it's part of the current widget tree.  Similarly, I assume that there's no way for _setCurrentMonth() to be called more than once in between builds - so we will not unnecessarily dispose/create page controllers?

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 15]
```diff
@@ -0,0 +1,1050 @@
...
+  AnimationController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = AnimationController(
+      value: widget.mode == DatePickerMode.year ? 1 : 0,
+      duration: const Duration(milliseconds: 200),
+      vsync: this,
+    )..addListener(() => setState(() {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Typically this is done by wrapping the `build()` widget with an AnimatedBuilder whose animation is _controller. In this case it looks like only the Transform widget needs to be wrapped with an AnimatedBuilder .. in fact maybe a RotationTransition would be better than an AnimatedBuilder

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 16]
```diff
@@ -0,0 +1,1050 @@
...
+                  onTap: widget.onTitlePressed,
+                  child: Padding(
+                    padding: const EdgeInsets.symmetric(horizontal: 8),
+                    child: Row(
+                      children: <Widget>[
+                        Flexible(
+                          child: Text(
+                            widget.title,
+                            overflow: TextOverflow.ellipsis,
+                            style: textTheme.subtitle2?.apply(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Do we really need to use 'apply' here? It's semantics are more complicated than copyWith

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+const Size _inputPortraitDialogSize = Size(330.0, 270.0);
+const Size _inputLandscapeDialogSize = Size(496, 160.0);
+const Duration _dialogSizeAnimationDuration = Duration(milliseconds: 200);
+
+/// Shows a dialog containing a Material Design date picker.
+///
+/// The returned [Future] resolves to the date selected by the user when the
+/// user confirms the dialog. If the user cancels the dialog, null is returned.
+///
+/// The [initialDate] is the initially shown and selected date in the picker.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: This sentence is a little hard to parse.

Maybe something like: When the date picker is first displayed, [initialDate] is selected.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 18]
```diff
@@ -0,0 +1,480 @@
...
+/// return true for [initialDate].
+///
+/// Optional strings for the [cancelText], [confirmText], [errorFormatText],
+/// [errorInvalidText], [fieldHintText], [fieldLabelText], and [helpText] allow
+/// you to override the default text used for various parts of the dialog:
+///
+///   * [cancelText], label on the cancel button.
+///   * [confirmText], label on the ok button.
+///   * [errorFormatText], message used when the input text isn't in a proper date format.
+///   * [errorInvalidText], message used when the input date isn't a selectable date.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: input date => input text

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart (hunk 10 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: There's no doc for the builder parameter.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+    final MaterialLocalizations localizations = MaterialLocalizations.of(context);
+    final Orientation orientation = MediaQuery.of(context).orientation;
+    final TextTheme textTheme = theme.textTheme;
+    final double textScaleFactor = math.min(MediaQuery.of(context).textScaleFactor, 1.3);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: This probably merits a brief explanation

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 20]
```diff
@@ -0,0 +1,480 @@
...
+
+    final String dateText = _selectedDate != null
+      ? localizations.formatMediumDate(_selectedDate)
+      // TODO(darrenaustin): localize 'Date'
+      : 'Date';
+    final Color dateColor = colorScheme.brightness == Brightness.light
+      ? colorScheme.onPrimary
+      : colorScheme.onSurface;
+    final TextStyle dateStyle = orientation == Orientation.landscape
+      ? textTheme.headline5?.apply(color: dateColor)
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: Same comment as elsewhere about apply instead of copyWith

**Review by HansMuller — COMMENTED** [review]
Mostly small stuff and wordsmithing.

File: packages/flutter/lib/src/material/material_localizations.dart (hunk 5 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material_localizations.dart]: Yeah, I will put that on my list and follow up later. There are other things like the DateTimeRange and other stuff in the date_util.dart that might be good candidates as well.

File: packages/flutter/lib/src/material/material_localizations.dart (hunk 6 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material_localizations.dart]: This is just a fallback for the default case.

Other locales are handled by the _MaterialLocalizationsDelegate:

https://github.com/flutter/flutter/pull/50546/files/4e3dc0117f2531952d92b702b2e09db2aa9f82c6#diff-945add11ecf717a6dad5cb70168cd894R625

Which will use the `intl.DateFormat.yMd` for the given locale. This is how the other formats that were already there were implemented.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart (hunk 8 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Not sure :). This was code from @rami-a that I didn't look that closely at. I'll spend some time with it and see if I can get better docs here and your next comment below.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart (hunk 11 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: Yup. I will do this in another pass after the date range picker gets in as well.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 21]
```diff
@@ -221,6 +221,29 @@ abstract class MaterialLocalizations {
...
+  String formatCompactDate(DateTime date);
+
+  /// Formats the date using a short-width format.
+  ///
+  /// Includes the abbreviation of the month, the day and year.
+  ///
+  /// Examples:
+  ///
+  /// - US English: Feb 21, 2019
+  /// - Russian: 21 февр. 2019 г.
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material_localizations.dart]: No, but I used the existing date formatters the the intl package to generate them, so I assumed it would be correct 😄. That said, I will ask around and see if anyone can review these.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 22]
```diff
@@ -434,6 +475,23 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
     'December',
   ];
 
+  /// Returns the number of days in a month, according to the proleptic
+  /// Gregorian calendar.
+  ///
+  /// This applies the leap year logic introduced by the Gregorian reforms of
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material_localizations.dart]: Yeah, this was from the original date_picker code. As Greg mentioned above, it might be good to add this to Dart's date package.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart (hunk 7 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: At one point I think I was wrapping the dialog with something else that got removed. I will simplify this. Thx.

File: packages/flutter/lib/src/material/material_localizations.dart (hunk 21 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/material/material_localizations.dart]: I'm a native Russian speaker and I approve this format :) (caveat: I'm not an expect on proper date formatting in Russia though)

[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/material/material_localizations.dart]: Maybe @mraleph can double-check.

[the remark this answers is not in the document]
Comment (mraleph) [reply packages/flutter/lib/src/material/material_localizations.dart]: Looks fine to me, though a tad bizarre. I thought abbreviation for February uses 3 letters rather than 4, but searching around reveals that people use 4... Well, I learned something today... Maybe I was asleep on that lesson in school :)

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+
+class _DateTextInputFormatter extends TextInputFormatter {
+
+  _DateTextInputFormatter(this.separator);
+
+  final String separator;
+
+  final WhitelistingTextInputFormatter _filterFormatter =
+    // Only allow digits and separators (slash, dot, comma, hyphen, space).
+    WhitelistingTextInputFormatter(RegExp(r'[\d\/\.,-\s]+'));
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/input_date_picker.dart]: There is already one above:

```dart
                // TODO(darrenaustin): localize date separator '/'
                _DateTextInputFormatter('/'),
```

These separators in the regex are any separators that the keyboard might be supplying that we always want to just map to the locale's separator. Probably not necessary, but I was worried about a date keyboard that comes up might not have locale's separator on it. The user would still be able to use these other separator keys and they would map to the right one.

File: packages/flutter/lib/src/material/pickers/date_utils.dart [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+  return DateTime(date.year, date.month, date.day);
+}
+
+/// Returns true if the two [DateTime] objects have the same day, month, and
+/// year.
+bool isSameDay(DateTime dateA, DateTime dateB) {
+  return
+    dateA.year == dateB.year &&
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_utils.dart]: Yeah, the `DateTime` constructor handles any wrapping needed for the day and month fields.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart (hunk 10 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_picker_dialog.dart]: See line 81 above.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart (hunk 14 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: Indeed. I just refactored the page controller out of the CalendarDatePicker and into the _MonthPicker so this will no longer be an issue.

**Comment by darrenaustin** [comment]
> Mostly small stuff and wordsmithing.

Thanks for the review @HansMuller. Can you take another look with my latest commits?

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
Do we know if this will address https://github.com/flutter/flutter/issues/49809 or https://github.com/flutter/flutter/issues/49810?

**Comment by darrenaustin** [comment]
> Do we know if this will address #49809 or #49810?

Unfortunately, this PR will not address those. They are still on my list. After I land the date range picker I will update them to address #49809. The Time Picker is currently being update by @rami-a and we can look at #49810 after he is finished with that.

**Comment by AndroidNils** [comment]
Thanks for this great PR, @darrenaustin! For our web app we are on master channel. Yesterday we pushed a new web app update and today we received the first user feedbacks that they love the new date picker 😍 
 
image

https://web.sharezone.net

**Comment by ilechcod** [comment]
Pls is the new material date picker available for use on flutter now? Do I need to update my current flutter version?

File: packages/flutter/lib/src/material/pickers/input_date_picker.dart [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+
+  @override
+  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
+    final TextEditingValue filteredValue = _filterFormatter.formatEditUpdate(oldValue, newValue);
+    return filteredValue.copyWith(
+      // Replace any separator character with the given separator
+      text: filteredValue.text.replaceAll(RegExp(r'[\D]'), separator),
+    );
+  }
+}
```
Comment (Hixie) [inline packages/flutter/lib/src/material/pickers/input_date_picker.dart:281]: @darrenaustin I'm about to check in something that replaces WhitelistingTextInputFormatter and BlacklistingTextInputFormatter with a single FilteringTextInputFormatter that handles replacement even for the case of allow-list filtering, which you might be able to use here.

File: packages/flutter/lib/src/material/pickers/date_picker_header.dart [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+
+import '../color_scheme.dart';
+import '../icon_button.dart';
+import '../text_theme.dart';
+import '../theme.dart';
+
+// NOTE: This is an internal implementation file. Even though there are public
+// classes and functions defined here, they are only meant to be used by the
+// date picker implementation and are not exported as part of the Material library.
+// See pickers.dart for exactly what is considered part of the public API.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/pickers/date_picker_header.dart:15]: As a general rule, we should just export everything and support everything as a public API. The reality is that everyone can see everything we do, so they'll just import the files directly if they want to, so we have to consider everything public anyway. I think we should just make everything public rather than start having semi-private APIs like this.

File: packages/flutter/lib/src/material/pickers/date_picker_header.dart [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+import '../color_scheme.dart';
+import '../icon_button.dart';
+import '../text_theme.dart';
+import '../theme.dart';
+
+// NOTE: This is an internal implementation file. Even though there are public
```
Comment (Hixie) [inline packages/flutter/lib/src/material/pickers/date_picker_header.dart:12]: trivial nit: "NOTE" is against our style guide. :-)
